### PR TITLE
Fix cleanup of Singletons and MixxxMainWindow

### PR DESCRIPTION
**Pull request by Holzhaus**
Here are some fixes I had to apply while working on a way to allow Mixxx to restart. Restarting Mixxx if off the table now, but IMHO it makes sense to apply the fixes anyway. I rebased on 2.2 so that this can go into 2.2.4.

### CODE REVIEW

**Comment by daschuer** [comment]
Does this fix a bug? It is hard to track which implications the finalize move has.

**Comment by Holzhaus** [comment]
> Does this fix a bug? It is hard to track which implications the finalize move has.

`finalize` deletes members of the mainwindow object. If the window would be shown again after close, Mixxx would crash. Likewise, if the window is destroyed without calling finalize, we'd leak memory. Both is currently is not possible, but deleting members should be done in the destructor.

**Comment by ywwg** [comment]
Can you explain why this needs to be backported to 2.2?  2.2 doesn't support restarting in-app

**Comment by Be-ing** [comment]
Changes should *only* be made to the stable branch to fix bugs or add controller mappings. Do not change the stable branch just to clean the code.

**Comment by Holzhaus** [comment]
Ok, I changed the base branch to `master`.

File: src/util/singleton.h [hunk 1]
```diff
@@ -14,15 +14,16 @@ class Singleton {
     }
 
     static T* instance() {
-        if (m_instance == NULL) {
-            qWarning() << "Singleton class has not been created yet, returning NULL";
+        if (m_instance == nullptr) {
```
Comment (uklotzde) [inline src/util/singleton.h:17]: `VERIFY_OR_DEBUG_ASSERT(m_instance)`? A developer should better notice whenever this happens. Not related to this PR, but why not improve the code if possible.

**Comment by uklotzde** [comment]
Works fine, thank you! LGTM

**Pull request merged**